services/distribution/src/main/java/app/coronawarn/server/services/distribution/assembly/component/MasterIndexFileCreator.java

```diff
@@ -0,0 +1,131 @@
+
+  private static final int DIRECTORY_SCANNING_MAX_DEPTH = 10;
+
+  private static final String NEW_LINE_SEPARATOR = "\r\n";
+
+  private final DistributionServiceConfig distributionServiceConfig;
+
+  private final OutputDirectoryProvider outputDirectoryProvider;
+
+  /**
+   * Creates a new instance with the given configuration and directory provider.
+   *
+   * @param distributionServiceConfig the distribution config.
+   * @param outputDirectoryProvider   the output directory provider.
+   */
+  public MasterIndexFileCreator(DistributionServiceConfig distributionServiceConfig,
+      OutputDirectoryProvider outputDirectoryProvider) {
+    this.distributionServiceConfig = distributionServiceConfig;
+    this.outputDirectoryProvider = outputDirectoryProvider;
+  }
+
+  /**
+   * Creates the master index file, and puts it in: version/_VERSION_/diagnosis-keys/index.
+   *
+   * @throws IOException in case the file could not be created
+   */
+  public void createIndex() throws IOException {
+    Path versionDirectory = getVersionDirectory();
+
+    try (Stream<Path> stream = Files.list(versionDirectory)) {
+      stream
+          .filter(Files::isDirectory)
+          .forEach(this::createIndexForCountriesAtVersionFolder);
+    }
+  }
+
+  private void createIndexForCountriesAtVersionFolder(Path versionFolder) {
+    var countryFolder = versionFolder
+        .resolve(distributionServiceConfig.getApi().getDiagnosisKeysPath())
+        .resolve(distributionServiceConfig.getApi().getCountryPath());
+
+    try (Stream<Path> stream = Files.list(countryFolder)) {
+      stream.filter(Files::isDirectory).forEach(this::generateMainIndex);
+    } catch (IOException e) {
+      throw new UncheckedIOException(e);
+    }
+  }
+
+  private void generateMainIndex(Path path) {
+    logger.info("Creating index for for {}", path.toAbsolutePath());
+
+    try (Stream<Path> stream = Files.walk(path, DIRECTORY_SCANNING_MAX_DEPTH)) {
+      String indexFileContent = stream
+          .filter(Files::isRegularFile)
+          .filter(MasterIndexFileCreator::isRelevantForMainIndex)
+          .map(Path::getParent)
+          .map(path::relativize)
+          .map(Path::toString)
+          .sorted()
+          .collect(Collectors.joining(NEW_LINE_SEPARATOR));
+
+      storeIndexFile(path, indexFileContent);
+    } catch (IOException e) {
+      throw new UncheckedIOException(e);
+    }
+  }
+
+  private void storeIndexFile(Path countryFolder, String content) {
+    Path targetIndexFile = countryFolder.resolve(distributionServiceConfig.getApi().getDiagnosisKeysIndexPath());
+
+    try {
+      logger.debug("Storing index file on {}", targetIndexFile.toAbsolutePath());
+
+      Files.writeString(targetIndexFile, content, StandardCharsets.UTF_8);
+    } catch (IOException e) {
+      throw new UncheckedIOException(e);
+    }
+  }
+
+  private static boolean isRelevantForMainIndex(Path path) {
+    return HOUR_INDEX_PATTERN.matcher(path.toString()).matches();
```

Comment:
Consider replacing a pattern matching on paths with an `instanceof`, checking if a file/directory is an instance of [DiagnosisKeysDateDirectory](https://github.com/corona-warn-app/cwa-server/blob/master/services/distribution/src/main/java/app/coronawarn/server/services/distribution/assembly/diagnosiskeys/structure/directory/DiagnosisKeysDateDirectory.java).